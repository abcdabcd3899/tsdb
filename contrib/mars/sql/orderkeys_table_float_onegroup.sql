\set tname test_orderkeys.t1_floattable_onegroup

-- this test is sensitive to cluster size, so force all the data to be stored
-- on seg0 only.
select gp_debug_set_create_table_default_numsegments(1);

-- create :tname with the first column "c1" as the group key
create table :tname (c1 float, c2 int)
 using mars
  with (global_order_col_= "{c1, c2}", local_order_col_= "{c1}")
 distributed by (c1)
;

insert into :tname select i / 4., i % 4 from generate_series(0, 15) i;

\i sql/orderkeys.template
