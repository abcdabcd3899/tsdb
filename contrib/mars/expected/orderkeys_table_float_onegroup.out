\set tname test_orderkeys.t1_floattable_onegroup
-- this test is sensitive to cluster size, so force all the data to be stored
-- on seg0 only.
select gp_debug_set_create_table_default_numsegments(1);
 gp_debug_set_create_table_default_numsegments 
-----------------------------------------------
 1
(1 row)

-- create :tname with the first column "c1" as the group key
create table :tname (c1 float, c2 int)
 using mars
  with (global_order_col_= "{c1, c2}", local_order_col_= "{c1}")
 distributed by (c1)
;
insert into :tname select i / 4., i % 4 from generate_series(0, 15) i;
\i sql/orderkeys.template
--
-- verify the grouping
--
select ctid, c1, c2 from :tname;
  ctid  |  c1  | c2 
--------+------+----
 (1,1)  |    0 |  0
 (1,2)  | 0.25 |  1
 (1,3)  |  0.5 |  2
 (1,4)  | 0.75 |  3
 (1,5)  |    1 |  0
 (1,6)  | 1.25 |  1
 (1,7)  |  1.5 |  2
 (1,8)  | 1.75 |  3
 (1,9)  |    2 |  0
 (1,10) | 2.25 |  1
 (1,11) |  2.5 |  2
 (1,12) | 2.75 |  3
 (1,13) |    3 |  0
 (1,14) | 3.25 |  1
 (1,15) |  3.5 |  2
 (1,16) | 3.75 |  3
(16 rows)

--
-- verify the fast seeking with only the first key
--
select c1, c2 from :tname where c1 <  1 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
  0.5 |  2
 0.75 |  3
(4 rows)

select c1, c2 from :tname where c1 <= 1 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
  0.5 |  2
 0.75 |  3
    1 |  0
(5 rows)

select c1, c2 from :tname where c1 =  1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
(1 row)

select c1, c2 from :tname where c1 >= 1 order by c1, c2;
  c1  | c2 
------+----
    1 |  0
 1.25 |  1
  1.5 |  2
 1.75 |  3
    2 |  0
 2.25 |  1
  2.5 |  2
 2.75 |  3
    3 |  0
 3.25 |  1
  3.5 |  2
 3.75 |  3
(12 rows)

select c1, c2 from :tname where c1 >  1 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
  1.5 |  2
 1.75 |  3
    2 |  0
 2.25 |  1
  2.5 |  2
 2.75 |  3
    3 |  0
 3.25 |  1
  3.5 |  2
 3.75 |  3
(11 rows)

--
-- verify the fast seeking with only the second key
--
select c1, c2 from :tname where c2 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  2 |  0
  3 |  0
(4 rows)

select c1, c2 from :tname where c2 <= 1 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
    1 |  0
 1.25 |  1
    2 |  0
 2.25 |  1
    3 |  0
 3.25 |  1
(8 rows)

select c1, c2 from :tname where c2 =  1 order by c1, c2;
  c1  | c2 
------+----
 0.25 |  1
 1.25 |  1
 2.25 |  1
 3.25 |  1
(4 rows)

select c1, c2 from :tname where c2 >= 1 order by c1, c2;
  c1  | c2 
------+----
 0.25 |  1
  0.5 |  2
 0.75 |  3
 1.25 |  1
  1.5 |  2
 1.75 |  3
 2.25 |  1
  2.5 |  2
 2.75 |  3
 3.25 |  1
  3.5 |  2
 3.75 |  3
(12 rows)

select c1, c2 from :tname where c2 >  1 order by c1, c2;
  c1  | c2 
------+----
  0.5 |  2
 0.75 |  3
  1.5 |  2
 1.75 |  3
  2.5 |  2
 2.75 |  3
  3.5 |  2
 3.75 |  3
(8 rows)

--
-- verify the fast seeking with both keys
--
select c1, c2 from :tname where c1 <  1 and c2 <  2 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
(2 rows)

select c1, c2 from :tname where c1 <  1 and c2 <= 2 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
  0.5 |  2
(3 rows)

select c1, c2 from :tname where c1 <  1 and c2 =  2 order by c1, c2;
 c1  | c2 
-----+----
 0.5 |  2
(1 row)

select c1, c2 from :tname where c1 <  1 and c2 >= 2 order by c1, c2;
  c1  | c2 
------+----
  0.5 |  2
 0.75 |  3
(2 rows)

select c1, c2 from :tname where c1 <  1 and c2 >  2 order by c1, c2;
  c1  | c2 
------+----
 0.75 |  3
(1 row)

select c1, c2 from :tname where c1 <= 1 and c2 <  2 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
    1 |  0
(3 rows)

select c1, c2 from :tname where c1 <= 1 and c2 <= 2 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
  0.5 |  2
    1 |  0
(4 rows)

select c1, c2 from :tname where c1 <= 1 and c2 =  2 order by c1, c2;
 c1  | c2 
-----+----
 0.5 |  2
(1 row)

select c1, c2 from :tname where c1 <= 1 and c2 >= 2 order by c1, c2;
  c1  | c2 
------+----
  0.5 |  2
 0.75 |  3
(2 rows)

select c1, c2 from :tname where c1 <= 1 and c2 >  2 order by c1, c2;
  c1  | c2 
------+----
 0.75 |  3
(1 row)

select c1, c2 from :tname where c1 =  1 and c2 <  2 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
(1 row)

select c1, c2 from :tname where c1 =  1 and c2 <= 2 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
(1 row)

select c1, c2 from :tname where c1 =  1 and c2 =  2 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 =  1 and c2 >= 2 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 =  1 and c2 >  2 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 >= 1 and c2 <  2 order by c1, c2;
  c1  | c2 
------+----
    1 |  0
 1.25 |  1
    2 |  0
 2.25 |  1
    3 |  0
 3.25 |  1
(6 rows)

select c1, c2 from :tname where c1 >= 1 and c2 <= 2 order by c1, c2;
  c1  | c2 
------+----
    1 |  0
 1.25 |  1
  1.5 |  2
    2 |  0
 2.25 |  1
  2.5 |  2
    3 |  0
 3.25 |  1
  3.5 |  2
(9 rows)

select c1, c2 from :tname where c1 >= 1 and c2 =  2 order by c1, c2;
 c1  | c2 
-----+----
 1.5 |  2
 2.5 |  2
 3.5 |  2
(3 rows)

select c1, c2 from :tname where c1 >= 1 and c2 >= 2 order by c1, c2;
  c1  | c2 
------+----
  1.5 |  2
 1.75 |  3
  2.5 |  2
 2.75 |  3
  3.5 |  2
 3.75 |  3
(6 rows)

select c1, c2 from :tname where c1 >= 1 and c2 >  2 order by c1, c2;
  c1  | c2 
------+----
 1.75 |  3
 2.75 |  3
 3.75 |  3
(3 rows)

select c1, c2 from :tname where c1 >  1 and c2 <  2 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
    2 |  0
 2.25 |  1
    3 |  0
 3.25 |  1
(5 rows)

select c1, c2 from :tname where c1 >  1 and c2 <= 2 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
  1.5 |  2
    2 |  0
 2.25 |  1
  2.5 |  2
    3 |  0
 3.25 |  1
  3.5 |  2
(8 rows)

select c1, c2 from :tname where c1 >  1 and c2 =  2 order by c1, c2;
 c1  | c2 
-----+----
 1.5 |  2
 2.5 |  2
 3.5 |  2
(3 rows)

select c1, c2 from :tname where c1 >  1 and c2 >= 2 order by c1, c2;
  c1  | c2 
------+----
  1.5 |  2
 1.75 |  3
  2.5 |  2
 2.75 |  3
  3.5 |  2
 3.75 |  3
(6 rows)

select c1, c2 from :tname where c1 >  1 and c2 >  2 order by c1, c2;
  c1  | c2 
------+----
 1.75 |  3
 2.75 |  3
 3.75 |  3
(3 rows)

--
-- verify the fast seeking with AND expr on the first key
--
select c1, c2 from :tname where c1 >= 1 and c1 <= 3 order by c1, c2;
  c1  | c2 
------+----
    1 |  0
 1.25 |  1
  1.5 |  2
 1.75 |  3
    2 |  0
 2.25 |  1
  2.5 |  2
 2.75 |  3
    3 |  0
(9 rows)

select c1, c2 from :tname where c1 >  1 and c1 <= 3 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
  1.5 |  2
 1.75 |  3
    2 |  0
 2.25 |  1
  2.5 |  2
 2.75 |  3
    3 |  0
(8 rows)

select c1, c2 from :tname where c1 >= 1 and c1 <  3 order by c1, c2;
  c1  | c2 
------+----
    1 |  0
 1.25 |  1
  1.5 |  2
 1.75 |  3
    2 |  0
 2.25 |  1
  2.5 |  2
 2.75 |  3
(8 rows)

select c1, c2 from :tname where c1 >  1 and c1 <  3 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
  1.5 |  2
 1.75 |  3
    2 |  0
 2.25 |  1
  2.5 |  2
 2.75 |  3
(7 rows)

--
-- verify the fast seeking with AND expr on the second key
--
select c1, c2 from :tname where c2 >= 1 and c2 <= 3 order by c1, c2;
  c1  | c2 
------+----
 0.25 |  1
  0.5 |  2
 0.75 |  3
 1.25 |  1
  1.5 |  2
 1.75 |  3
 2.25 |  1
  2.5 |  2
 2.75 |  3
 3.25 |  1
  3.5 |  2
 3.75 |  3
(12 rows)

select c1, c2 from :tname where c2 >  1 and c2 <= 3 order by c1, c2;
  c1  | c2 
------+----
  0.5 |  2
 0.75 |  3
  1.5 |  2
 1.75 |  3
  2.5 |  2
 2.75 |  3
  3.5 |  2
 3.75 |  3
(8 rows)

select c1, c2 from :tname where c2 >= 1 and c2 <  3 order by c1, c2;
  c1  | c2 
------+----
 0.25 |  1
  0.5 |  2
 1.25 |  1
  1.5 |  2
 2.25 |  1
  2.5 |  2
 3.25 |  1
  3.5 |  2
(8 rows)

select c1, c2 from :tname where c2 >  1 and c2 <  3 order by c1, c2;
 c1  | c2 
-----+----
 0.5 |  2
 1.5 |  2
 2.5 |  2
 3.5 |  2
(4 rows)

--
-- verify the fast seeking with AND expr on both keys
--
select c1, c2 from :tname where c1 >= 1 and c1 <= 3 and c2 >= 1 and c2 <= 3 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
  1.5 |  2
 1.75 |  3
 2.25 |  1
  2.5 |  2
 2.75 |  3
(6 rows)

select c1, c2 from :tname where c1 >  1 and c1 <= 3 and c2 >  1 and c2 <= 3 order by c1, c2;
  c1  | c2 
------+----
  1.5 |  2
 1.75 |  3
  2.5 |  2
 2.75 |  3
(4 rows)

select c1, c2 from :tname where c1 >= 1 and c1 <  3 and c2 >= 1 and c2 <  3 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
  1.5 |  2
 2.25 |  1
  2.5 |  2
(4 rows)

select c1, c2 from :tname where c1 >  1 and c1 <  3 and c2 >  1 and c2 <  3 order by c1, c2;
 c1  | c2 
-----+----
 1.5 |  2
 2.5 |  2
(2 rows)

select c1, c2 from :tname where c1 =  1 and c2 >= 1 and c2 <= 3 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 =  1 and c2 >  1 and c2 <= 3 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 =  1 and c2 >= 1 and c2 <  3 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 =  1 and c2 >  1 and c2 <  3 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 >= 1 and c1 <= 3 and c2 =  1 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
 2.25 |  1
(2 rows)

select c1, c2 from :tname where c1 >  1 and c1 <= 3 and c2 =  1 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
 2.25 |  1
(2 rows)

select c1, c2 from :tname where c1 >= 1 and c1 <  3 and c2 =  1 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
 2.25 |  1
(2 rows)

select c1, c2 from :tname where c1 >  1 and c1 <  3 and c2 =  1 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
 2.25 |  1
(2 rows)

--
-- verify the IN expr
--
select c1, c2 from :tname where c1 in ( 0, 1, 3 ) order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  3 |  0
(3 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
(1 row)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 <= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
(2 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 =  1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
(1 row)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 >= 1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  3 |  0
(2 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 >  1 order by c1, c2;
 c1 | c2 
----+----
  3 |  0
(1 row)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  3 |  0
(3 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 <= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  3 |  0
(3 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 =  1 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 >= 1 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 >  1 order by c1, c2;
 c1 | c2 
----+----
(0 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
 0.75 |  3
    1 |  0
 1.25 |  1
 1.75 |  3
    2 |  0
 2.25 |  1
 2.75 |  3
    3 |  0
 3.25 |  1
 3.75 |  3
(12 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 <  1 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
 0.75 |  3
(3 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 <= 1 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
 0.75 |  3
    1 |  0
(4 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 =  1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
(1 row)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 >= 1 order by c1, c2;
  c1  | c2 
------+----
    1 |  0
 1.25 |  1
 1.75 |  3
    2 |  0
 2.25 |  1
 2.75 |  3
    3 |  0
 3.25 |  1
 3.75 |  3
(9 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 >  1 order by c1, c2;
  c1  | c2 
------+----
 1.25 |  1
 1.75 |  3
    2 |  0
 2.25 |  1
 2.75 |  3
    3 |  0
 3.25 |  1
 3.75 |  3
(8 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  2 |  0
  3 |  0
(4 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 <= 1 order by c1, c2;
  c1  | c2 
------+----
    0 |  0
 0.25 |  1
    1 |  0
 1.25 |  1
    2 |  0
 2.25 |  1
    3 |  0
 3.25 |  1
(8 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 =  1 order by c1, c2;
  c1  | c2 
------+----
 0.25 |  1
 1.25 |  1
 2.25 |  1
 3.25 |  1
(4 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 >= 1 order by c1, c2;
  c1  | c2 
------+----
 0.25 |  1
 0.75 |  3
 1.25 |  1
 1.75 |  3
 2.25 |  1
 2.75 |  3
 3.25 |  1
 3.75 |  3
(8 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 >  1 order by c1, c2;
  c1  | c2 
------+----
 0.75 |  3
 1.75 |  3
 2.75 |  3
 3.75 |  3
(4 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 in ( 0, 1, 3 ) order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  3 |  0
(3 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 2 ) and c1 in ( 1, 2, 3 ) order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  2 |  0
(2 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 2 ) and c2 in ( 1, 2, 3 ) order by c1, c2;
  c1  | c2 
------+----
 0.25 |  1
  0.5 |  2
 1.25 |  1
  1.5 |  2
 2.25 |  1
  2.5 |  2
 3.25 |  1
  3.5 |  2
(8 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 2 ) and c1 in ( 1, 2, 3 )
                       and c2 in ( 0, 1, 2 ) and c2 in ( 1, 2, 3 ) order by c1, c2;
 c1 | c2 
----+----
(0 rows)

